sessionctx/variable/sysvar_test.go

```diff
@@ -87,3 +88,83 @@ func (*testSysVarSuite) TestError(c *C) {
 		c.Assert(terror.ToSQLError(err).Code != mysql.ErrUnknown, IsTrue)
 	}
 }
+
+func (*testSysVarSuite) TestRegistrationOfNewSysVar(c *C) {
+	count := len(GetSysVars())
+	sv := SysVar{Scope: ScopeGlobal | ScopeSession, Name: "mynewsysvar", Value: On, Type: TypeBool, SetSession: func(s *SessionVars, val string) error {
+		return fmt.Errorf("set should fail")
+	}}
+
+	RegisterSysVar(&sv)
+	c.Assert(count+1, Equals, len(GetSysVars()))
+
+	sysVar := GetSysVar("mynewsysvar")
+	c.Assert(sysVar, NotNil)
+
+	vars := NewSessionVars()
+
+	// It is a boolean, try to set it to a bogus value
+	_, err := sysVar.Validate(vars, "ABCD", ScopeSession)
+	c.Assert(err, NotNil)
+
+	// Boolean oN or 1 converts to canonical ON or OFF
+	normalizedVal, err := sysVar.Validate(vars, "oN", ScopeSession)
+	c.Assert(normalizedVal, Equals, "ON")
+	normalizedVal, err = sysVar.Validate(vars, "0", ScopeSession)
+	c.Assert(normalizedVal, Equals, "OFF")
+
+	err = sysVar.SetSessionFromHook(vars, "OFF") // default is on
+	c.Assert(err.Error(), Matches, "set should fail")
+
+	// Test unregistration restores previous count
+	UnregisterSysVar("mynewsysvar")
+	c.Assert(count, Equals, len(GetSysVars()))
+}
+
+func (*testSysVarSuite) TestIntValidation(c *C) {
+	sv := SysVar{Scope: ScopeGlobal | ScopeSession, Name: "mynewsysvar", Value: "123", Type: TypeInt, MinValue: 10, MaxValue: 300, AllowAutoValue: true}
+	vars := NewSessionVars()
+
+	_, err := sv.Validate(vars, "oN", ScopeSession)
+	c.Assert(err.Error(), Equals, "[variable:1232]Incorrect argument type to variable 'mynewsysvar'")
+
+	_, err = sv.Validate(vars, "301", ScopeSession)
+	c.Assert(err.Error(), Equals, "[variable:1231]Variable 'mynewsysvar' can't be set to the value of '301'")
+
+	_, err = sv.Validate(vars, "5", ScopeSession)
+	c.Assert(err.Error(), Equals, "[variable:1231]Variable 'mynewsysvar' can't be set to the value of '5'")
+
+	val, err := sv.Validate(vars, "300", ScopeSession)
+	c.Assert(err, IsNil)
+	c.Assert(val, Equals, "300")
+
+	// out of range but permitted due to auto value
```

Comment:
just a question, what happens if `-1` is a valid value and `AllowAutoValue` is true?

Reply:
```
func (sv *SysVar) checkInt64SystemVar(value string, vars *SessionVars) (string, error) {
	if sv.AllowAutoValue && value == "-1" {
		return value, nil
	}
```
Looks like it doesn't matter. `-1` will be returned if it is valid or `AllowAutoValue == true`. But you can not use `-1` to identify "auto" anymore.

Reply:
Yeah, if either is supported the value is permitted.

I think in most cases that use an AutoValue, it refers to something like worker threads, so it is likely always a positive value.

But the autovalue when assigned to the sysvar means nothing by itself. It is whatever reads the value that needs to also support the idea of an auto value.

Reply:
AutoValue is a MySQL concept btw. It is used for a lot of settings in performance_schema for example. We are just copying the behavior for some of our own settings.